runtime/src/slices.rs

```diff
@@ -0,0 +1,236 @@
+use alloc::boxed::Box;
+use core::marker::PhantomData;
+use core::ops::{Deref, DerefMut};
+
+/// This is equivalent to &[T], except it has a stable repr(C) layout
+#[repr(C)]
+pub struct DiplomatSlice<'a, T> {
+    // Invariant: ptr is a valid ptr to the beginning of an &[T] allocation
```

Comment:
I think this needs work around empty slices. Rust doesn't require an allocation (even an empty one) for empty slices, and Diplomat currently allows passing `NULL, 0`.